services/pull/protection_check.go

```diff
@@ -0,0 +1,86 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package pull
+
+import (
+	"context"
+
+	"code.gitea.io/gitea/models"
+	user_model "code.gitea.io/gitea/models/user"
+	asymkey_service "code.gitea.io/gitea/services/asymkey"
+)
+
+// CheckPullProtection check if the pull mergable based on all conditions (branch protection, merge options, ...)
+func CheckPullProtection(ctx context.Context, doer *user_model.User, perm *models.Permission, pr *models.PullRequest, manuallMerge, force bool) error {
```

Comment:
I think the name should be changed. It doesn't only check protection, but also other situation. Maybe `CheckPullMergable` is better.